src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -320,74 +179,67 @@ bool AdaptDispatch::_CursorMovePosition(const std::optional<size_t> row, const s
 
     if (success)
     {
-        // handle optional parameters. If not specified, keep same cursor position from what we just loaded.
-        size_t rowActual = 0;
-        size_t columnActual = 0;
-
-        if (row)
+        // Calculate the viewport boundaries as inclusive values.
+        // srWindow is exclusive so we need to subtract 1 from the bottom.
+        const int viewportTop = csbiex.srWindow.Top;
+        const int viewportBottom = csbiex.srWindow.Bottom - 1;
+
+        // Calculate the absolute margins of the scrolling area.
+        const int topMargin = viewportTop + _scrollMargins.Top;
+        const int bottomMargin = viewportTop + _scrollMargins.Bottom;
+        const bool marginsSet = topMargin < bottomMargin;
+
+        // For relative movement, the given offsets will be relative to
+        // the current cursor position.
+        int row = csbiex.dwCursorPosition.Y;
+        int col = csbiex.dwCursorPosition.X;
+
+        // But if the row is absolute, it will be relative to the top of the
+        // viewport, or the top margin, depending on the origin mode.
+        if (rowOffset.IsAbsolute)
         {
-            if (row.value() != 0)
-            {
-                rowActual = row.value() - 1; // In VT, the origin is 1,1. For our array, it's 0,0. So subtract 1.
-
-                // If the origin mode is relative, and the scrolling region is set (the bottom is non-zero),
-                // line numbers start at the top margin of the scrolling region, and cannot move below the bottom.
-                if (_isOriginModeRelative && _scrollMargins.Bottom != 0)
-                {
-                    rowActual += _scrollMargins.Top;
-                    rowActual = std::min<decltype(rowActual)>(rowActual, _scrollMargins.Bottom);
-                }
-            }
-            else
-            {
-                success = false; // The parser should never return 0 (0 maps to 1), so this is a failure condition.
-            }
+            row = _isOriginModeRelative ? topMargin : viewportTop;
         }
-        else
+
+        // And if the column is absolute, it'll be relative to column 0.
+        // Horizontal positions are not affected by the viewport.
+        if (colOffset.IsAbsolute)
         {
-            rowActual = csbiex.dwCursorPosition.Y - csbiex.srWindow.Top; // remember, in VT speak, this is relative to the viewport. not absolute.
+            col = 0;
         }
 
-        if (column)
+        // Adjust the base position by the given offsets and clamp the results.
+        // The row is constrained within the viewport's vertical boundaries,
+        // while the column is constrained by the buffer width.
+        row = std::clamp(row + rowOffset.Value, viewportTop, viewportBottom);
+        col = std::clamp(col + colOffset.Value, 0, csbiex.dwSize.X - 1);
```

Comment:
There's some `Clamp` provisions on `Viewport` that could bring the row/col pair inside a set of bounds, presuming you can set up the `Viewport` to meet your requirements. If you use it to make the above math more explicit, you might already have one for nice `.Clamp()`ing here.

If `Viewport` doesn't quite do what you want, I'd like to hear how we could perhaps make it better or why it was/is bad.

Reply:
This sounds really promising, but there are couple of issues that make it not workable at the moment. The first is that the area we're dealing with is a combination of two sources - the vertical extent of the visible viewport, and the horizontal extent of the buffer - and we can't easily create a `Viewport` from that.

That said, that structure is something we use quite a lot in the VT code, so it may worth adding a method that returned the "active viewport", and then refactor all the code to make use of that structure at some point. I think that would also allow us to make better use of `_virtualBottom`, which I think isn't always taken into account correclty (that's on my TODO list of things to check).

The second issue is that the `Viewport` coordinates are all currently short ints, but for this code to work correctly, we need to do the calculations with a wider range than that, and only convert down to a short after the values have been clamped. If we had to narrow the coordinates before clamping, it's not going to work.

But hopefully this could eventually be addressed by the changes you're proposing in #4015.

Reply:
I've been playing around with this some more, and another approach I tried was to have two separate `Viewport` variables - one from the `srWindow` for the vertical extent, and another from the `dwSize` for the horizontal extent. That still then required additional `ClampX` and `ClampY` methods in the `Viewport` (taking an int) to clamp the row and column individually.

The generated assembly with this approach was not great, though, because there's a bunch of overhead associated with the `SMALL_RECT` structs which aren't needed, as well as inefficiencies from working with shorts. So I think I'd still be inclined to leave the code as is, and consider refactoring later when we have more efficient `Viewport` or `Size` abstractions that better match our needs.

Reply:
Works for me. Thank you for the careful consideration.